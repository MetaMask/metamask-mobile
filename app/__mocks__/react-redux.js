export * from 'react-redux-test';
