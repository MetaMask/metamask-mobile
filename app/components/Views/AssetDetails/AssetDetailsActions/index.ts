export { default, AssetDetailsActionsRef } from './AssetDetailsActions';
