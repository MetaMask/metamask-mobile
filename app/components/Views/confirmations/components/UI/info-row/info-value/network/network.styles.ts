import { StyleSheet } from 'react-native';

import { Theme } from '../../../../../../../../util/theme/models';
import { fontStyles } from '../../../../../../../../styles/common';

const styleSheet = (params: { theme: Theme }) => {
  const { theme } = params;

  return StyleSheet.create({
    container: {
      display: 'flex',
      flexDirection: 'row',
      alignItems: 'center',
    },
    value: {
      color: theme.colors.text.default,
      ...fontStyles.normal,
      fontSize: 14,
      marginLeft: 8,
    },
  });
};

export default styleSheet;
