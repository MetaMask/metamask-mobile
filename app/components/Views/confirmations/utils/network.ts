import { Hex, isCaipChainId } from '@metamask/utils';
import {
  isTestNet,
  getTestNetImageByChainId,
  getDefaultNetworkByChainId,
} from '../../../../util/networks';
import {
  UnpopularNetworkList,
  CustomNetworkImgMapping,
  PopularList,
  getNonEvmNetworkImageSourceByChainId,
} from '../../../../util/networks/customNetworks';

export function getNetworkBadgeSource(chainId: Hex) {
  if (isTestNet(chainId)) return getTestNetImageByChainId(chainId);
  const defaultNetwork = getDefaultNetworkByChainId(chainId) as
    | {
        imageSource: string;
      }
    | undefined;

  if (defaultNetwork) {
    return defaultNetwork.imageSource;
  }

  const unpopularNetwork = UnpopularNetworkList.find(
    (networkConfig) => networkConfig.chainId === chainId,
  );

  const customNetworkImg = CustomNetworkImgMapping[chainId];

  const popularNetwork = PopularList.find(
    (networkConfig) => networkConfig.chainId === chainId,
  );

  const network = unpopularNetwork || popularNetwork;
  if (network) {
    return network.rpcPrefs.imageSource;
  }
  if (isCaipChainId(chainId)) {
    return getNonEvmNetworkImageSourceByChainId(chainId);
  }
  if (customNetworkImg) {
    return customNetworkImg;
  }
}
