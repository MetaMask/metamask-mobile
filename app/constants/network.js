export const MAINNET = 'mainnet';
export const HOMESTEAD = 'homestead';
export const ROPSTEN = 'ropsten';
export const KOVAN = 'kovan';
export const RINKEBY = 'rinkeby';
export const GOERLI = 'goerli';
export const RPC = 'rpc';
export const NO_RPC_BLOCK_EXPLORER = 'NO_BLOCK_EXPLORER';
export const PRIVATENETWORK = 'PRIVATENETWORK';
export const DEFAULT_MAINNET_CUSTOM_NAME = 'Ethereum Main Custom';

/**
 * @enum {string}
 */
export const NETWORKS_CHAIN_ID = {
  MAINNET: '1',
  OPTIMISM: '10',
  KOVAN: '42',
  BSC: '56',
  POLYGON: '137',
  FANTOM: '250',
  ARBITRUM: '42161',
  AVAXCCHAIN: '43114',
  CELO: '42220',
  HARMONY: '1666600000',
};
