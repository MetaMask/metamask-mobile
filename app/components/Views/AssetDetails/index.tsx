import React, { useEffect, useMemo } from 'react';
import {
  View,
  StyleSheet,
  ScrollView,
  Text,
  TouchableOpacity,
  InteractionManager,
} from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { getNetworkNavbarOptions } from '../../UI/Navbar';
import { fontStyles } from '../../../styles/common';
import ClipboardManager from '../../../core/ClipboardManager';
import { showAlert } from '../../../actions/alert';
import { strings } from '../../../../locales/i18n';
import { useDispatch, useSelector } from 'react-redux';
import EthereumAddress from '../../UI/EthereumAddress';
import Icon from 'react-native-vector-icons/Feather';
import TokenImage from '../../UI/TokenImage';
import Networks, { getDecimalChainId } from '../../../util/networks';
import Engine from '../../../core/Engine';
import Logger from '../../../util/Logger';
import NotificationManager from '../../../core/NotificationManager';
import AppConstants from '../../../core/AppConstants';
import { Token as TokenType } from '@metamask/assets-controllers';
import {
  balanceToFiat,
  renderFromTokenMinimalUnit,
} from '../../../util/number';
import WarningMessage from '../SendFlow/WarningMessage';
import { useTheme } from '../../../util/theme';
import { MetaMetricsEvents } from '../../../core/Analytics';
import AnalyticsV2 from '../../../util/analyticsV2';
import Routes from '../../../constants/navigation/Routes';

const createStyles = (colors: any) =>
  StyleSheet.create({
    container: {
      padding: 16,
      backgroundColor: colors.background.default,
      alignItems: 'flex-start',
    },
    descriptionContainer: {
      marginTop: 4,
      flexDirection: 'row',
      alignItems: 'center',
    },
    tokenImage: { height: 36, width: 36, marginRight: 8 },
    sectionTitleLabel: {
      ...(fontStyles.bold as any),
      fontSize: 16,
      color: colors.text.default,
      marginTop: 32,
    },
    firstSectionTitle: {
      marginTop: 8,
    },
    descriptionLabel: {
      ...(fontStyles.normal as any),
      fontSize: 16,
      color: colors.text.default,
    },
    hideButton: {
      marginTop: 48,
    },
    hideButtonLabel: {
      ...(fontStyles.normal as any),
      fontSize: 16,
      color: colors.error.default,
    },
    addressLinkLabel: {
      ...(fontStyles.normal as any),
      fontSize: 16,
      color: colors.primary.default,
    },
    copyIcon: {
      marginLeft: 4,
      color: colors.primary.default,
    },
    warningBanner: { marginTop: 8 },
    warningBannerDesc: { color: colors.text.default },
    warningBannerLink: { color: colors.primary.default },
  });

interface Props {
  route: {
    params: {
      address: string;
    };
  };
}

const AssetDetails = (props: Props) => {
  const { address } = props.route.params;
  const { colors } = useTheme();
  const styles = createStyles(colors);
  const navigation = useNavigation();
  const dispatch = useDispatch();
  const network = useSelector(
    (state: any) => state.engine.backgroundState.NetworkController,
  );
  const tokens = useSelector(
    (state: any) =>
      state.engine.backgroundState.TokensController.tokens as TokenType[],
  );
  const conversionRate = useSelector(
    (state: any) =>
      state.engine.backgroundState.CurrencyRateController.conversionRate,
  );
  const currentCurrency = useSelector(
    (state: any) =>
      state.engine.backgroundState.CurrencyRateController.currentCurrency,
  );
  const primaryCurrency = useSelector(
    (state: any) => state.settings.primaryCurrency,
  );
  const tokenBalances = useSelector(
    (state: any) =>
      state.engine.backgroundState.TokenBalancesController.contractBalances,
  );
  const tokenExchangeRates = useSelector(
    (state: any) =>
      state.engine.backgroundState.TokenRatesController.contractExchangeRates,
  );
  const token = useMemo(
    () => tokens.find((rawToken) => rawToken.address === address),
    [tokens, address],
  );
  const { symbol, decimals, aggregators = [] } = token as TokenType;

  const getNetworkName = () => {
    let name = '';
    if (network.provider.nickname) {
      name = network.provider.nickname;
    } else {
      name =
        (Networks as any)[network.provider.type]?.name ||
        { ...Networks.rpc, color: null }.name;
    }
    return name;
  };

  useEffect(() => {
    navigation.setOptions(
      getNetworkNavbarOptions(
        'Token Details',
        false,
        navigation,
        colors,
        undefined,
        true,
      ),
    );
  }, [navigation, colors]);

  const copyAddressToClipboard = async () => {
    await ClipboardManager.setString(address);
    dispatch(
      showAlert({
        isVisible: true,
        autodismiss: 1500,
        content: 'clipboard-alert',
        data: { msg: strings('detected_tokens.address_copied_to_clipboard') },
      }),
    );
  };

  const triggerHideToken = () => {
    const { TokensController, NetworkController } = Engine.context as any;
    navigation.navigate(Routes.MODAL.ROOT_MODAL_FLOW, {
      screen: 'AssetHideConfirmation',
      params: {
        onConfirm: () => {
          navigation.navigate('WalletView');
          InteractionManager.runAfterInteractions(async () => {
            try {
              await TokensController.ignoreTokens([address]);
              NotificationManager.showSimpleNotification({
                status: `simple_notification`,
                duration: 5000,
                title: strings('wallet.token_toast.token_hidden_title'),
                description: strings('wallet.token_toast.token_hidden_desc', {
                  tokenSymbol: symbol,
                }),
              });
              AnalyticsV2.trackEvent(MetaMetricsEvents.TOKENS_HIDDEN, {
                location: 'token_details',
                token_standard: 'ERC20',
                asset_type: 'token',
                tokens: [`${symbol} - ${address}`],
                chain_id: getDecimalChainId(
                  NetworkController?.state?.provider?.chainId,
                ),
              });
            } catch (err) {
              Logger.log(err, 'AssetDetails: Failed to hide token!');
            }
          });
        },
      },
    });
  };

  const renderWarningBanner = () => (
    <WarningMessage
      style={styles.warningBanner}
      warningMessage={
        <>
          <Text style={styles.warningBannerDesc}>
            {strings('asset_overview.were_unable')} {symbol}{' '}
            {strings('asset_overview.balance')}{' '}
            <Text
              suppressHighlighting
              onPress={() => {
                navigation.navigate('Webview', {
                  screen: 'SimpleWebview',
                  params: {
                    url: AppConstants.URLS.TOKEN_BALANCE,
                    title: strings('asset_overview.troubleshoot'),
                  },
                });
              }}
              style={styles.warningBannerLink}
            >
              {strings('asset_overview.troubleshooting_missing')}{' '}
            </Text>
            {strings('asset_overview.for_help')}
          </Text>
        </>
      }
    />
  );

  const renderSectionTitle = (title: string, isFirst?: boolean) => (
    <Text
      style={[styles.sectionTitleLabel, isFirst && styles.firstSectionTitle]}
    >
      {title}
    </Text>
  );

  const renderSectionDescription = (description: string) => (
    <Text style={[styles.descriptionLabel, styles.descriptionContainer]}>
      {description}
    </Text>
  );

  const renderTokenSymbol = () => (
    <View style={styles.descriptionContainer}>
      <TokenImage
        asset={{ address }}
        containerStyle={styles.tokenImage}
        iconStyle={styles.tokenImage}
      />
      <Text style={styles.descriptionLabel}>{symbol}</Text>
    </View>
  );

  const renderTokenBalance = () => {
    let balanceDisplay = '';
    const exchangeRate =
      address in tokenExchangeRates ? tokenExchangeRates[address] : undefined;
    const balance =
      address in tokenBalances
        ? renderFromTokenMinimalUnit(tokenBalances[address], decimals)
        : undefined;
    const balanceFiat = balance
      ? balanceToFiat(balance, conversionRate, exchangeRate, currentCurrency)
      : undefined;

    if (balance === undefined && balanceFiat === undefined) {
      // Couldn't load balance
      return renderWarningBanner();
    }

    if (primaryCurrency === 'ETH') {
      balanceDisplay = balanceFiat
        ? `${balance} (${balanceFiat})`
        : `${balance}`;
    } else {
      balanceDisplay = balanceFiat
        ? `${balanceFiat} (${balance})`
        : `${balance}`;
    }

    return (
      <View style={styles.descriptionContainer}>
        <Text style={styles.descriptionLabel}>{balanceDisplay}</Text>
      </View>
    );
  };

  const renderHideButton = () => (
    <TouchableOpacity
      onPress={triggerHideToken}
      hitSlop={{ top: 24, bottom: 24, left: 24, right: 24 }}
      style={styles.hideButton}
    >
      <Text style={styles.hideButtonLabel}>
        {strings('asset_details.hide_cta')}
      </Text>
    </TouchableOpacity>
  );

  const renderTokenAddressLink = () => (
    <TouchableOpacity
      hitSlop={{ top: 24, bottom: 24, left: 24, right: 24 }}
      onPress={copyAddressToClipboard}
      style={styles.descriptionContainer}
    >
      <EthereumAddress
        style={styles.addressLinkLabel}
        address={address}
        type={'short'}
      />
      <Icon style={styles.copyIcon} name={'copy'} size={16} />
    </TouchableOpacity>
  );

  return (
    <ScrollView contentContainerStyle={styles.container}>
      {renderSectionTitle(strings('asset_details.token'), true)}
      {renderTokenSymbol()}
      {renderSectionTitle(strings('asset_details.amount'))}
      {renderTokenBalance()}
      {renderSectionTitle(strings('asset_details.address'))}
      {renderTokenAddressLink()}
      {renderSectionTitle(strings('asset_details.decimal'))}
      {renderSectionDescription(String(decimals))}
      {renderSectionTitle(strings('asset_details.network'))}
      {renderSectionDescription(getNetworkName())}
      {renderSectionTitle(strings('asset_details.lists'))}
      {renderSectionDescription(aggregators.join(', '))}
      {renderHideButton()}
    </ScrollView>
  );
};

export default AssetDetails;
