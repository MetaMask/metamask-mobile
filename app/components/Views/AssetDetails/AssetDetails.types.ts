import { Hex } from '@metamask/utils';
import { TokenI } from '../../UI/Tokens/types';

export type AssetDetailsParams = {
  address: Hex;
  chainId: Hex;
  asset: TokenI;
};
