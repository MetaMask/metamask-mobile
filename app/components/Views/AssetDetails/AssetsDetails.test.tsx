import React from 'react';
import AssetDetails from '.';
import configureMockStore from 'redux-mock-store';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import { CHAIN_IDS } from '@metamask/transaction-controller';
import { backgroundState } from '../../../util/test/initial-root-state';
import { mockNetworkState } from '../../../util/test/network';
import Engine from '../../../core/Engine';

jest.mock('../../../core/ClipboardManager', () => ({
  setString: jest.fn(),
}));

jest.mock('../../../util/networks', () => ({
  ...jest.requireActual('../../../util/networks'),
  getDecimalChainId: jest.fn(() => 1),
}));

jest.mock('@react-navigation/compat', () => ({
  withNavigation: (Component: React.ComponentType) => Component,
}));

const MOCK_ADDRESS_1 = '0x0';

jest.mock('../../../core/Engine', () => ({
  context: {
    TokensController: {
      ignoreTokens: jest.fn(),
    },
    NetworkController: {
      findNetworkClientIdByChainId: jest.fn(() => 'mainnet'),
    },
  },
}));

jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn(() => ({
    setOptions: jest.fn(),
    navigate: jest.fn(),
  })),
}));

jest.mock('../../../core/NotificationManager', () => ({
  showSimpleNotification: jest.fn(),
}));

const mockStore = configureMockStore();

const initialState = {
  engine: {
    backgroundState: {
      ...backgroundState,
      NetworkController: {
        ...mockNetworkState({
          chainId: CHAIN_IDS.MAINNET,
          id: 'mainnet',
          nickname: 'Ethereum Mainnet',
          ticker: 'ETH',
        }),
      },
      TokenBalancesController: {
        tokenBalances: {
          [MOCK_ADDRESS_1]: {
            '0x1': {
              '0xAddress': '0xde0b6B3A7640000',
            },
          },
        },
      },
      TokensController: {
        tokens: [
          {
            address: '0xAddress',
            symbol: 'TKN',
            decimals: 18,
            aggregators: ['Metamask', 'CMC'],
          },
        ],
      },
      AccountsController: {
        internalAccounts: {
          selectedAccount: 'id',
          accounts: {
            id: {
              address: MOCK_ADDRESS_1,
            },
          },
        },
      },
      TokenRatesController: {
        contractExchangeRates: {
          '0xAddress': { price: 2 },
        },
        marketData: {
          '0xAddress': { price: 2 },
        },
      },
      CurrencyRateController: {
        conversionRate: 1,
        currentCurrency: 'USD',
      },
    },
  },
  settings: {
    primaryCurrency: 'ETH',
  },
};

const store = mockStore(initialState);

describe('AssetDetails', () => {
  const renderComponent = () =>
    render(
      <Provider store={store}>
        <AssetDetails
          route={{
            params: {
              address: '0xAddress',
            },
          }}
        />
      </Provider>,
    );

  it('renders correctly', () => {
    const { toJSON } = renderComponent();
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders token details correctly', () => {
    const { getByText } = renderComponent();

    expect(getByText('Token')).toBeDefined();
    expect(getByText('Token Amount')).toBeDefined();
    expect(getByText('Token contract address')).toBeDefined();
    expect(getByText('Token decimal')).toBeDefined();
    expect(getByText('Network')).toBeDefined();
    expect(getByText('Token Lists')).toBeDefined();
  });

  it('copies address to clipboard and shows alert', async () => {
    const { getByText } = render(
      <Provider store={store}>
        <AssetDetails
          route={{
            params: {
              address: '0xAddress',
            },
          }}
        />
      </Provider>,
    );

    const copyElement = getByText('0xAddr...ress');
    fireEvent.press(copyElement);

    await waitFor(() => {
      expect(
        Engine.context.TokensController.ignoreTokens,
      ).not.toHaveBeenCalled();
    });
  });

  it('navigates to asset hide confirmation on pressing hide button', async () => {
    const { getByText } = renderComponent();

    fireEvent.press(getByText('Hide token'));

    await waitFor(() => {
      expect(
        Engine.context.TokensController.ignoreTokens,
      ).not.toHaveBeenCalled();
    });
  });

  it('renders warning banner if balance is undefined', () => {
    const mockEmptyState = {
      ...initialState,
      engine: {
        backgroundState: {
          ...initialState.engine.backgroundState,
          TokenBalancesController: {
            tokenBalances: {},
          },
        },
      },
    };

    const mockStoreEmpty = mockStore(mockEmptyState);

    const { getByText } = render(
      <Provider store={mockStoreEmpty}>
        <AssetDetails
          route={{
            params: {
              address: '0xAddress',
            },
          }}
        />
      </Provider>,
    );

    expect(getByText('troubleshooting missing balances')).toBeDefined();
  });

  it('renders balance in ETH if primary currency is ETH', () => {
    const { getByText } = renderComponent();
    expect(getByText('1')).toBeDefined();
  });

  it('renders balance in fiat if primary currency is fiat', () => {
    const mockFiatState = {
      ...initialState,
      settings: {
        primaryCurrency: 'fiat',
      },
    };
    const mockStoreFiat = mockStore(mockFiatState);

    const { getByText } = render(
      <Provider store={mockStoreFiat}>
        <AssetDetails
          route={{
            params: {
              address: '0xAddress',
            },
          }}
        />
      </Provider>,
    );

    expect(getByText('1')).toBeDefined();
  });
});
