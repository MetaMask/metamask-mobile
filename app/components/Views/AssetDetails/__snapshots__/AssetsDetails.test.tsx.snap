// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetDetails renders correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "alignItems": "flex-start",
      "backgroundColor": "#ffffff",
      "padding": 16,
    }
  }
>
  <View>
    <Text
      style={
        [
          {
            "color": "#141618",
            "fontFamily": "EuclidCircularB-Bold",
            "fontSize": 16,
            "fontWeight": "600",
            "marginTop": 32,
          },
          {
            "marginTop": 8,
          },
        ]
      }
    >
      Token
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "marginTop": 4,
        }
      }
    >
      <View
        style={
          [
            {
              "height": 50,
              "width": 50,
            },
            {
              "height": 36,
              "marginRight": 8,
              "width": 36,
            },
            {
              "borderRadius": 25,
              "overflow": "hidden",
            },
          ]
        }
      />
      <Text
        style={
          {
            "color": "#141618",
            "fontFamily": "EuclidCircularB-Regular",
            "fontSize": 16,
            "fontWeight": "400",
          }
        }
      >
        TKN
      </Text>
    </View>
    <Text
      style={
        [
          {
            "color": "#141618",
            "fontFamily": "EuclidCircularB-Bold",
            "fontSize": 16,
            "fontWeight": "600",
            "marginTop": 32,
          },
          undefined,
        ]
      }
    >
      Token Amount
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "marginTop": 4,
        }
      }
    >
      <Text
        style={
          {
            "color": "#141618",
            "fontFamily": "EuclidCircularB-Regular",
            "fontSize": 16,
            "fontWeight": "400",
          }
        }
      >
        1
      </Text>
    </View>
    <Text
      style={
        [
          {
            "color": "#141618",
            "fontFamily": "EuclidCircularB-Bold",
            "fontSize": 16,
            "fontWeight": "600",
            "marginTop": 32,
          },
          undefined,
        ]
      }
    >
      Token contract address
    </Text>
    <TouchableOpacity
      hitSlop={
        {
          "bottom": 24,
          "left": 24,
          "right": 24,
          "top": 24,
        }
      }
      onPress={[Function]}
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "marginTop": 4,
        }
      }
    >
      <Text
        numberOfLines={1}
        style={
          {
            "color": "#0376c9",
            "fontFamily": "EuclidCircularB-Regular",
            "fontSize": 16,
            "fontWeight": "400",
          }
        }
      >
        0xAddr...ress
      </Text>
      <Text
        allowFontScaling={false}
        style={
          [
            {
              "color": undefined,
              "fontSize": 16,
            },
            {
              "color": "#0376c9",
              "marginLeft": 4,
            },
            {
              "fontFamily": "Feather",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        î¥„
      </Text>
    </TouchableOpacity>
    <Text
      style={
        [
          {
            "color": "#141618",
            "fontFamily": "EuclidCircularB-Bold",
            "fontSize": 16,
            "fontWeight": "600",
            "marginTop": 32,
          },
          undefined,
        ]
      }
    >
      Token decimal
    </Text>
    <Text
      style={
        [
          {
            "color": "#141618",
            "fontFamily": "EuclidCircularB-Regular",
            "fontSize": 16,
            "fontWeight": "400",
          },
          {
            "alignItems": "center",
            "flexDirection": "row",
            "marginTop": 4,
          },
        ]
      }
    >
      18
    </Text>
    <Text
      style={
        [
          {
            "color": "#141618",
            "fontFamily": "EuclidCircularB-Bold",
            "fontSize": 16,
            "fontWeight": "600",
            "marginTop": 32,
          },
          undefined,
        ]
      }
    >
      Network
    </Text>
    <Text
      style={
        [
          {
            "color": "#141618",
            "fontFamily": "EuclidCircularB-Regular",
            "fontSize": 16,
            "fontWeight": "400",
          },
          {
            "alignItems": "center",
            "flexDirection": "row",
            "marginTop": 4,
          },
        ]
      }
    >
      Ethereum Mainnet
    </Text>
    <Text
      style={
        [
          {
            "color": "#141618",
            "fontFamily": "EuclidCircularB-Bold",
            "fontSize": 16,
            "fontWeight": "600",
            "marginTop": 32,
          },
          undefined,
        ]
      }
    >
      Token Lists
    </Text>
    <Text
      style={
        [
          {
            "color": "#141618",
            "fontFamily": "EuclidCircularB-Regular",
            "fontSize": 16,
            "fontWeight": "400",
          },
          {
            "alignItems": "center",
            "flexDirection": "row",
            "marginTop": 4,
          },
        ]
      }
    >
      Metamask, CMC
    </Text>
    <TouchableOpacity
      hitSlop={
        {
          "bottom": 24,
          "left": 24,
          "right": 24,
          "top": 24,
        }
      }
      onPress={[Function]}
      style={
        {
          "marginTop": 48,
        }
      }
    >
      <Text
        style={
          {
            "color": "#d73847",
            "fontFamily": "EuclidCircularB-Regular",
            "fontSize": 16,
            "fontWeight": "400",
          }
        }
      >
        Hide token
      </Text>
    </TouchableOpacity>
  </View>
</RCTScrollView>
`;
